### PR TITLE
Update the time period based on timestamp proto

### PR DESCRIPTION
**Pull request by songchen2022**
### Changes proposed in this Pull Request:

Update the start/end date in the TimePeriod message following google.protobuf.Timestamp.
The reason why we update is to keep update with latest content API version and deprecate field that previously used in unstable version.
API will still be backward compatible so there is no worries for merchants who has installed the old version.
[https://screenshot.googleplex.com/7essYuaBVYnACqV](url)
[https://screenshot.googleplex.com/3CmCh48JqiYfCPH](url)

### CODE REVIEW

File: src/Coupon/WCCouponAdapter.php [hunk 1]
```diff
@@ -184,7 +184,7 @@ protected function get_wc_coupon_effective_dates( WC_Coupon $wc_coupon ): ?strin
 			$end_date = clone $start_date;
 			$end_date->add( new DateInterval( 'PT1S' ) );
 		}
-		return sprintf( '%s/%s', (string) $start_date, (string) $end_date );
+		return sprintf( '{"startTime": "%s", "endTime": "%s"}', (string) $start_date, (string) $end_date );
```
Comment (chennyxie) [inline src/Coupon/WCCouponAdapter.php]: Content API has new field to accept promo effective period.
 this change changes the string format, but does not return the data and put it into the new field?

Comment (songchen2022) [reply src/Coupon/WCCouponAdapter.php]: Thanks for your comment. Have updated this part to put the data into the time period field.

File: composer.json [hunk 2]
```diff
@@ -48,7 +48,11 @@
 		"platform": {
 			"php": "7.3.27"
 		},
-		"sort-packages": true
+		"sort-packages": true,
+		"allow-plugins": {
```
Comment (mikkamp) [inline composer.json:52]: Can we exclude these changes from this PR, as it's also bumping package versions in the composer.lock file. We have a separate PR to handle this already #1180
I think it's best to leave the changes to one PR so we can confirm all the package version bumping doesn't cause any issues.

File: package.json [hunk 3]
```diff
@@ -89,6 +89,7 @@
 		"@wordpress/url": "^2.22.2",
 		"classnames": "^2.3.1",
 		"gridicons": "^3.3.1",
+		"joi": "^17.6.0",
```
Comment (mikkamp) [inline package.json:92]: Is there a particular reason this extra package (and the package-lock.json) needs to be included in this PR?

**Review by mikkamp — COMMENTED** [review]
I mainly wanted to leave some comments about parts which can be excluded from this PR, and also wanted to note that there are some alignment / spacing issues (which will cause the automated tests to fail). You might want to run the code changes through phpcs to get those fixed.

I haven't tested the actual changes but wanted to mention that they rely on the library updates in #1204 (before then TimePeriod is not available). Those changes will be released by the time this PR is ready so it shouldn't be an issue.

**Pull request closed without merging**